paymentgateway/paypal/tests/paymentgateway_paypal_ipn_test.php

```diff
@@ -0,0 +1,237 @@
+        $ex->userid = 3;
+        $ex->payment_status = 'Completed';
+        $ex->pending_reason = null;
+        $this->assertEquals($ex, $data);
+
+        // Similar IPN but with no null values
+        $post = $this->generate_simulated_ipn('ipn_normal2');
+        $ipn = new ipn();
+        $data = $ipn->process_ipn($post);
+
+        // Expected result.
+        $ex = new stdClass();
+        $ex->txn_type = 'cart';
+        $ex->business = 'test@business.example.com';
+        $ex->charset = 'UTF-8';
+        $ex->parent_txn_id = '1QR097329Y527661V';
+        $ex->receiver_email = 'test@business.example.com';
+        $ex->receiver_id = 'T9CHS2ZUN6BL6';
+        $ex->resend = 'true';
+        $ex->residence_country = 'AU';
+        $ex->test_ipn = '1';
+        $ex->txn_id = '1QR097329Y527661V';
+        $ex->first_name = 'John';
+        $ex->last_name = 'Doe';
+        $ex->payer_id = 'ZFDYPPYELC4KS';
+        $ex->item_name1 = 'Test course_1+2=3';
+        $ex->mc_currency = 'USD';
+        $ex->mc_gross = '50.00';
+        $ex->payment_date = '16:09:34 Sep 14, 2020 PDT';
+        $ex->courseid = 2;
+        $ex->userid = 3;
+        $ex->payment_status = 'Pending';
+        $ex->pending_reason = 'echeck';
+        $this->assertEquals($ex, $data);
+    }
+
+    // Test processing of IPN without custom value.
+    public function not_a_test_yet() {
+        $generator = $this->getDataGenerator();
+
+        $courserecord = new \stdClass();
+        $courserecord->fullname = 'Test course_1+2=3';
+        $courserecord->idnumber = 10;
+        $course = $generator->create_course($courserecord);
+         
+        $userrecord = new \stdClass();
+        $userrecord->idnumber = 25;
+        $user = $generator->create_user($userrecord);
+
+        $this->setUser($user);
+    }
+
+    // Test full purchase
+    public function test_valid_purchase() {
+        $this->resetAfterTest();
+        global $DB;
+
+        // Catch fake paypal ipn
+        $post = $this->generate_simulated_ipn('ipn_normal');
+        $ipn = new ipn();
+        $data = $ipn->process_ipn($post);
+        $this->assertEquals($ipn->validate($data), 'INVALID');
+
+        $files = ['ipn_normal', 'ipn_normal2', 'ipn_failed'];
+        for ($t = 0; $t < count($files); $t++) {
+            $course = $this->getDataGenerator()->create_course();
+            $user = $this->getDataGenerator()->create_user();
+            $DB->insert_record('enrol', array('enrol' => 'payment', 'courseid' => $course->id));
+    
+            $post = $this->generate_simulated_ipn($files[$t]);
+            $ipn = new ipn();
+            $data = $ipn->process_ipn($post);
+            $data->courseid = $course->id;
+            $data->userid = $user->id;
+            $result_of_test = $ipn->submit_data('VERIFIED', $data);
+            // Get Purchase Proof
+            $record2 = $DB->get_record_sql('SELECT * FROM {tool_paymentplugin_purchases} as ppp JOIN {paymentgateway_paypal} as 
```

Comment:
This should be formatted with Moodle SQL formatting: https://docs.moodle.org/dev/SQL_coding_style